File: solutions/typescript/split-second-stopwatch/1/split-second-stopwatch.ts
export class SplitSecondStopwatch {
  private _state: 'ready' | 'running' | 'stopped';
  private _currentTime: number; // in seconds
  private _previousLaps: number[];
  private _startTime: number | null;

  constructor() {
    this._state = 'ready';
    this._currentTime = 0;
    this._previousLaps = [];
    this._startTime = null;
  }

  public get state(): 'ready' | 'running' | 'stopped' {
    return this._state;
  }

  public get currentLap(): string {
    let current = this._currentTime;
    if (this._state === 'running' && this._startTime !== null) {
      current += (Date.now() - this._startTime) / 1000;
    }
    return this._formatTime(current);
  }

  public get total(): string {
    let total = this._previousLaps.reduce((sum, lap) => sum + lap, 0);
    total += this._currentTime;
    if (this._state === 'running' && this._startTime !== null) {
      total += (Date.now() - this._startTime) / 1000;
    }
    return this._formatTime(total);
  }

  public get previousLaps(): string[] {
    return this._previousLaps.map(lap => this._formatTime(lap));
  }

  public start(): SplitSecondStopwatch {
    if (this._state === 'running') {
      throw new Error('cannot start an already running stopwatch');
    }
    if (this._state === 'ready' || this._state === 'stopped') {
      this._state = 'running';
      this._startTime = Date.now();
    }
    return this;
  }

  public stop(): SplitSecondStopwatch {
    if (this._state !== 'running') {
      throw new Error('cannot stop a stopwatch that is not running');
    }
    if (this._state === 'running' && this._startTime !== null) {
      this._currentTime += (Date.now() - this._startTime) / 1000;
      this._state = 'stopped';
      this._startTime = null;
    }
    return this;
  }

  public lap(): SplitSecondStopwatch {
    if (this._state !== 'running') {
      throw new Error('cannot lap a stopwatch that is not running');
    }
    if (this._state === 'running' && this._startTime !== null) {
      const elapsed = (Date.now() - this._startTime) / 1000;
      const lapTime = this._currentTime + elapsed;
      this._previousLaps.push(lapTime);
      this._currentTime = 0;
      this._startTime = Date.now();
    }
    return this;
  }

  public reset(): SplitSecondStopwatch {
    if (this._state !== 'stopped') {
      throw new Error('cannot reset a stopwatch that is not stopped');
    }
    this._state = 'ready';
    this._currentTime = 0;
    this._previousLaps = [];
    this._startTime = null;
    return this;
  }

  public advanceTime(duration: string | number): SplitSecondStopwatch {
    // Parse time string like "00:00:05" to seconds
    let seconds: number;
    if (typeof duration === 'string') {
      const parts = duration.split(':').map(Number);
      if (parts.length === 3) {
        seconds = parts[0] * 3600 + parts[1] * 60 + parts[2];
      } else {
        seconds = parseFloat(duration);
      }
    } else {
      seconds = duration;
    }
    
    if (this._state === 'running') {
      this._currentTime += seconds;
    }
    return this;
  }

  private _formatTime(seconds: number): string {
    if (isNaN(seconds) || seconds === undefined || seconds === null) {
      return '00:00:00';
    }
    
    const totalSeconds = Math.floor(seconds);
    const hours = Math.floor(totalSeconds / 3600);
    const minutes = Math.floor((totalSeconds % 3600) / 60);
    const secs = totalSeconds % 60;

    return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
  }
}

File: solutions/java/split-second-stopwatch/1/src/main/java/SplitSecondStopwatch.java
import java.util.ArrayList;
import java.util.List;

public class SplitSecondStopwatch {
    private enum State { READY, RUNNING, STOPPED }
    
    private State state = State.READY;
    private long currentTime = 0;
    private long lapStartTime = 0;
    
    // ⭐ NEW: Track the frozen duration of current lap when stopped
    private long currentLapDuration = 0;
    
    private long accumulatedTime = 0;
    private final List<Long> previousLapTimes = new ArrayList<>();
    
    public void start() {
        if (state == State.RUNNING) {
            throw new IllegalStateException("cannot start an already running stopwatch");
        }
        if (state == State.READY || state == State.STOPPED) {
            lapStartTime = currentTime;
            state = State.RUNNING;
        }
    }

    public void stop() {
        if (state != State.RUNNING) {
            throw new IllegalStateException("cannot stop a stopwatch that is not running");
        }
        // ⭐ FIX: Freeze current lap duration when stopping
        currentLapDuration = currentTime - lapStartTime;
        accumulatedTime += currentLapDuration;
        state = State.STOPPED;
    }

    public void reset() {
        if (state != State.STOPPED) {
            throw new IllegalStateException("cannot reset a stopwatch that is not stopped");
        }
        state = State.READY;
        accumulatedTime = 0;
        currentLapDuration = 0;  // ⭐ FIX: Clear frozen lap duration
        lapStartTime = 0;
        previousLapTimes.clear();
    }

    public void lap() {
        if (state != State.RUNNING) {
            throw new IllegalStateException("cannot lap a stopwatch that is not running");
        }
        long lapDuration = currentTime - lapStartTime;
        previousLapTimes.add(lapDuration);
        accumulatedTime += lapDuration;
        // ⭐ FIX: Reset for new lap segment
        currentLapDuration = 0;
        lapStartTime = currentTime;
    }

    public String state() {
        return state.name().toLowerCase();
    }

    public String currentLap() {
        // ⭐ FIX: State-aware calculation
        if (state == State.RUNNING) {
            // Running: frozen duration + elapsed since lap start
            return formatTime(currentLapDuration + (currentTime - lapStartTime));
        } else {
            // Stopped/Ready: just the frozen duration
            return formatTime(currentLapDuration);
        }
    }

    public String total() {
        if (state == State.RUNNING) {
            return formatTime(accumulatedTime + (currentTime - lapStartTime));
        } else {
            return formatTime(accumulatedTime);
        }
    }

    public List<String> previousLaps() {
        List<String> result = new ArrayList<>();
        for (long lapTime : previousLapTimes) {
            result.add(formatTime(lapTime));
        }
        return result;
    }

    public void advanceTime(String timeString) {
        currentTime += parseTime(timeString);
    }
    
    private long parseTime(String timeString) {
        String[] parts = timeString.split(":");
        long hours = Long.parseLong(parts[0]);
        long minutes = Long.parseLong(parts[1]);
        long seconds = Long.parseLong(parts[2]);
        return hours * 3600 + minutes * 60 + seconds;
    }
    
    private String formatTime(long totalSeconds) {
        long hours = totalSeconds / 3600;
        long remaining = totalSeconds % 3600;
        long minutes = remaining / 60;
        long seconds = remaining % 60;
        return String.format("%02d:%02d:%02d", hours, minutes, seconds);
    }
}